helix-core/src/syntax.rs

```diff
@@ -84,6 +86,32 @@ pub struct IndentQuery {
     pub outdent: HashSet<String>,
 }
 
+#[derive(Debug)]
+pub struct TextObjectQuery {
+    pub query: Query,
+}
+
+impl TextObjectQuery {
+    /// Run the query on the given node and return sub nodes which match given
+    /// capture ("function.inside", "class.around", etc).
+    pub fn capture_nodes<'a>(
+        &'a self,
+        capture_name: &str,
+        node: Node<'a>,
+        slice: RopeSlice<'a>,
+        cursor: &'a mut QueryCursor,
+    ) -> Option<impl Iterator<Item = Node<'a>>> {
```

Comment:
I quite like this returning an iterator 👍

Reply:
Ah wait, can't you fully replace your code with https://docs.rs/tree-sitter/0.20.0/tree_sitter/struct.QueryMatch.html#method.nodes_for_capture_index?

Reply:
We still have to manually get the capture index and the captures, and `nodes_for_capture_index()` only replaces [the condition for the filter map](https://github.com/helix-editor/helix/pull/728/files#diff-e78218ec4fe8cd27d4141d6d1077b4e929f734f03e97dcadfcc82f03d9ed661cR109). `nodes_for_capture_index()`also returns an iterator of matches, so we will have to `flat_map` it:

```rust
        let captures = cursor.matches(&self.query, node, RopeProvider(slice));

        captures
            .flat_map(move |ref mat| mat.nodes_for_capture_index(capture_idx).collect::<Vec<_>>())
            .into()
```

The Vec allocation is required because of lifetime issues, so we end up having to allocate for every single match on the whole tree.